### PR TITLE
WYSIWYG Field type core contribution

### PR DESCRIPTION
**Pull request by urdhvatech**
Created a new field type that uses latest tinyMCE to have HTML Editor in any module.
Admin can create field type called WYSIWYG in studio.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Final checklist

- [X] My code follows the code style of this project found [here](https://docs.suitecrm.com/community/contributing-code/coding-standards/).
- [X] My change requires a change to the documentation.
- [X] I have read the [**How to Contribute**](https://docs.suitecrm.com/community/contributing-code/) guidelines.

### CODE REVIEW

**Comment by samus-aran** [comment]
Thanks @urdhvatech for the core contribution! I'm sure the community will be keen to get this in! :heart:

**Comment by horus68** [comment]
To me: Self notification for the language change!

File: include/Smarty/plugins/modifier.htmlparser.php [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+/**
+ * Smarty HTML parser modifier plugin
+ *
+ * Type:     modifier<br>
+ * Name:     htmlparser<br>
+ * Purpose:  parse html to display
+ * @param string
+ * @return string
+ */
+function smarty_modifier_htmlparser($string, $uc_digits = false)
```
Comment (Abuelodelanada) [inline include/Smarty/plugins/modifier.htmlparser.php]: You are not using the variable `$uc_digits` in the function.

File: modules/DynamicFields/templates/Fields/Forms/wysiwyg.php [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+ * In accordance with Section 7(b) of the GNU Affero General Public License version 3,
+ * these Appropriate Legal Notices must retain the display of the "Powered by
+ * SugarCRM" logo and "Supercharged by SuiteCRM" logo. If the display of the logos is not
+ * reasonably feasible for technical reasons, the Appropriate Legal Notices must
+ * display the words "Powered by SugarCRM" and "Supercharged by SuiteCRM".
+ *
+ * This file was contributed by Urdhva tech private limited <contact@urdhva-tech.com>
+ **/
+if(!defined('sugarEntry') || !sugarEntry) die('Not A Valid Entry Point');
+function get_body(&$ss, $vardef) {
```
Comment (Abuelodelanada) [inline modules/DynamicFields/templates/Fields/Forms/wysiwyg.php]: You are not using `$vardef` variable in the function

**Comment by pribeiro42** [comment]
Thanks a lot for this enhancement. But shouldn't the license headers refer to GNU AGPL instead of the Sugar Professional Subscription Agreement ?

**Comment by Abuelodelanada** [comment]
Wonderful contribution @urdhvatech !!

**Comment by connorshea** [comment]
Is this also compatible with #4511?

**Comment by urdhvatech** [comment]
> 
> 
> Wonderful contribution @urdhvatech !!

Thanks @Abuelodelanada

File: include/Smarty/plugins/modifier.htmlparser.php (hunk 1 above)
Comment (urdhvatech) [reply include/Smarty/plugins/modifier.htmlparser.php]: Thanks, I have just committed the changes.

File: modules/DynamicFields/templates/Fields/Forms/wysiwyg.php (hunk 2 above)
Comment (urdhvatech) [reply modules/DynamicFields/templates/Fields/Forms/wysiwyg.php]: Thanks, I have just committed the changes.

**Comment by urdhvatech** [comment]
> Is this also compatible with #4511?

We have used the latest available library(4.3.4) in SuiteCRM at https://github.com/salesagility/SuiteCRM/blob/master/include/javascript/mozaik/vendor/tinymce/tinymce/

**Comment by samus-aran** [comment]
> Thanks a lot for this enhancement. But shouldn't the license headers refer to GNU AGPL instead of the Sugar Professional Subscription Agreement ?

Yes, that was addressed. We've confirmed that this code is not part of professional but a developer error (copy & paste is a terrible thing sometimes).

**Comment by cameronblaikie** [comment]
Assessed :+1:

**Comment by samus-aran** [comment]
Travis is failing @urdhvatech 
Do you run the unit and acceptance locally?
Error found is below:
Time: 5.06 minutes, Memory: 20.00MB

There was 1 failure:

---------
1) [35;1mElasticsearchCest:[39;22m Test search founds
[37;41;1m Test [39;49;22m tests/acceptance/lib/Search/ElasticsearchCest.php:testSearchFounds
[37;41;1m Step [39;49;22m See "Create Account",".actionmenulink"
[37;41;1m Fail [39;49;22m Element located either by name, CSS or XPath element with '.actionmenulink' was not found.
[33m
Scenario Steps:
[39m
[1m 11. $I->see("Create Account",".actionmenulink")[22m at [32mtests/_support/Step/Acceptance/AccountsTester.php:33[39m
 10. $I->click("Accounts","#toolbar.desktop-toolbar  > ul.nav.navbar-nav > ...") at [32mtests/_support/Step/Acceptance/NavigationBarTester.php:96[39m
 9. $I->waitForElementVisible("#toolbar.desktop-toolbar  > ul.nav.navba...",120) at [32mtests/_support/Step/Acceptance/NavigationBarTester.php:95[39m
 8. $I->click("All","#toolbar.desktop-toolbar  > ul.nav.navbar-nav > li.top...") at [32mtests/_support/Step/Acceptance/NavigationBarTester.php:93[39m
 7. $I->wait(1) at [32mtests/_support/Step/Acceptance/NavigationBarTester.php:92[39m
 6. $I->waitForElementVisible("#toolbar.desktop-toolbar  > ul.nav.navbar...",30) at [32mtests/_support/Step/Acceptance/NavigationBarTester.php:91[39m

**Comment by cameronblaikie** [comment]
Assessed :+1:

File: include/generic/SugarWidgets/SugarWidgetFieldwysiwyg.php [hunk 3]
```diff
@@ -0,0 +1,51 @@
+<?php
+if (!defined('sugarEntry') || !sugarEntry) {
```
Comment (code-ph0y) [inline include/generic/SugarWidgets/SugarWidgetFieldwysiwyg.php]: Please put this under the copyright statement

File: include/SugarFields/Fields/Wysiwyg/SugarFieldWysiwyg.php [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+  function getEditViewSmarty($parentFieldArray, $vardef, $displayParams, $tabindex) {
+        $this->setup($parentFieldArray, $vardef, $displayParams, $tabindex);
+      require_once('include/SugarTinyMCE.php');
+      global $json;
+    
+    if(empty($json)) {
+      $json = getJSONobj();
+    }
+    $form_name = '';
+    if(!empty($this->ss->_tpl_vars['displayParams']['formName']))
```
Comment (code-ph0y) [inline include/SugarFields/Fields/Wysiwyg/SugarFieldWysiwyg.php]: Please use if statements with open and closing curly braces 
`
if (true) {
    logic
} else {
    logic
}
`
Coding standard can be viewed at https://www.php-fig.org/psr/psr-2/

File: modules/DynamicFields/templates/Fields/TemplateWysiwyg.php [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+ * SugarCRM" logo and "Supercharged by SuiteCRM" logo. If the display of the logos is not
+ * reasonably feasible for technical reasons, the Appropriate Legal Notices must
+ * display the words "Powered by SugarCRM" and "Supercharged by SuiteCRM".
+ *
+ * This file was contributed by Urdhva tech private limited <contact@urdhva-tech.com>
+ **/
+if(!defined('sugarEntry') || !sugarEntry) die('Not A Valid Entry Point');
+// NOTE => field type
+require_once('modules/DynamicFields/templates/Fields/TemplateField.php');
+class TemplateWysiwyg extends TemplateField{
```
Comment (code-ph0y) [inline modules/DynamicFields/templates/Fields/TemplateWysiwyg.php]: Please fix code formatting

class opening curry brace should be on a new line: 

``` 
class TemplateWysiwyg extends TemplateField
{
     ....
```

File: modules/DynamicFields/templates/Fields/TemplateWysiwyg.php [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+	var $type = 'wysiwyg';
+	var $len = '';
+    var $inline_edit = 0;
```
Comment (code-ph0y) [inline modules/DynamicFields/templates/Fields/TemplateWysiwyg.php]: Fix code formatting.

indenting should be 4 spaces.

Example: 
```
class TemplateWysiwyg extends TemplateField
{
    var $type = 'wysiwyg';
    var $len = '';
    var $inline_edit = 0;
...
```

File: modules/DynamicFields/templates/Fields/TemplateWysiwyg.php [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+	function get_field_def(){
```
Comment (code-ph0y) [inline modules/DynamicFields/templates/Fields/TemplateWysiwyg.php]: Please add public, protected, or private to functions. (Encapsulation).

Furthermore, put opening curly braces of functions on new line.
Example:
``` 
public function get_field_def()
{
    ...

File: modules/DynamicFields/templates/Fields/TemplateWysiwyg.php [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+
```
Comment (code-ph0y) [inline modules/DynamicFields/templates/Fields/TemplateWysiwyg.php]: Fix code formatting.

indenting should be 4 spaces.

File: modules/AOS_PDF_Templates/templateParser.php [hunk 9]
```diff
@@ -59,54 +66,51 @@ public function parse_template_bean($string, $key, &$focus)
             if (isset($field_def['name']) && $field_def['name'] != '') {
                 $fieldName = $field_def['name'];
                 if ($field_def['type'] == 'currency') {
-                    $repl_arr[$key . "_" . $fieldName] = currency_format_number($focus->$fieldName, $params = array('currency_symbol' => false));
-                } else {
-                    if (($field_def['type'] == 'radioenum' || $field_def['type'] == 'enum' || $field_def['type'] == 'dynamicenum') && isset($field_def['options'])) {
-                        $repl_arr[$key . "_" . $fieldName] = translate($field_def['options'], $focus->module_dir, $focus->$fieldName);
+                    $repl_arr[$key . "_" . $fieldName] = currency_format_number($focus->$fieldName,
```
Comment (code-ph0y) [inline modules/AOS_PDF_Templates/templateParser.php]: Please fix code formatting:

If using multiple line functions - use the following format:
```
    $repl_arr[$key . "_" . $fieldName] = currency_format_number(
         $focus->$fieldName,
         $params = array(
             'currency_symbol' => false
         )
    );
```

File: modules/DynamicFields/templates/Fields/TemplateWysiwyg.php (hunk 6 above)
Comment (Abuelodelanada) [reply modules/DynamicFields/templates/Fields/TemplateWysiwyg.php]: Besides, Visibility MUST be declared on all properties [to follow PSR2](
https://www.php-fig.org/psr/psr-2/#42-properties)

```php
<?php
namespace Vendor\Package;

class ClassName
{
    public $foo = null;
}

```

**Comment by cameronblaikie** [comment]
Assessed :+1:

**Comment by aureh12** [comment]
If we force type='wysiwyg' on a text field in EditView metadata, tinyMCE is not displayed.
Javascript error : tinyMCE is not defined.

**Comment by aureh12** [comment]
If we use Extension to change the type of a text field, the new wysiwyg is not working also.
$dictionary['XXX']['fields']['description']['type']='wysiwyg';
$dictionary['XXX']['fields']['description']['dbtype']='text';
Then tinyMCE doesn't display on Edit view.

**Comment by IBlasterus** [comment]
SuiteCRM version 7.11.15.
Field type _wysiwyg_ doesn't work.
I added this field in Studio, but in EditView this field looks like a just TextArea without TinyMCE.

**Comment by pgorod** [comment]
@urdhvatech do you understand why this error is happening?

https://community.suitecrm.com/t/how-to-convert-default-description-field-into-wysiwyg-editor-for-task-module/74368

**Comment by urdhvatech** [comment]
@IBlasterus I tried to generate the issue on same version,and it works for me. Can you mention exact steps?

@pgorod Replied, In that case you also have to mention dbtype too.

**Comment by IBlasterus** [comment]
@urdhvatech
1. Administration - Studio - Contacts - create field - type wysiwyg.
Screenshot_1
2. Place it on the View and EditView.
Screenshot_2
3. Trying edit the field in module Contacts, but can't see TinyMCE.
Screenshot_3

**Comment by urdhvatech** [comment]
@IBlasterus 
Do you see any error in your console?
Also can you try login with english language and try?

**Comment by IBlasterus** [comment]
@urdhvatech 
Thanks for hint.
I found problem.
I am getting console error:
```
index.php:61 GET http://example.com/suitecrm/include/javascript/mozaik/vendor/tinymce/tinymce/tinymce.min.js?v=nnC9G512Ez28yOCjc43I0Q net::ERR_ABORTED 404 (Not Found)
Uncaught ReferenceError: tinyMCE is not defined
    at index.php:1395
```
Because of file [composer.json](https://github.com/salesagility/SuiteCRM/blob/master/composer.json) doesn't have this dependence.

**Comment by IBlasterus** [comment]
@urdhvatech 
Thanks for hint.
I solved the problem when replace folder **include/javascript/mozaik** from [your repository](https://github.com/salesagility/SuiteCRM/tree/master/include/javascript/mozaik).
Screenshot_1

I not change this folder. Just [download zip from official page](https://suitecrm.com/download/).

**Pull request merged**